### ISSUE TITLE
how to train a network that doesn't require any training data

### ISSUE THREAD

**Issue by jeff-regier** [PyTorchLightning/pytorch-lightning#2310]
The Wake-Sleep algorithm doesn't require any data during the sleep phase (effectively it generates it's own data). pytorch-lightning, however, appears to require a `train_dataloader()` method.

The only way I have to make pytorch-lightning run at all (for this admitted unusual case) is to specify some dummy dataset in `train_dataloader`, and then to ignore the data that gets passed to `training_step`. But I don't like that cycles are spent iterating through irrelevant data then. Is there a more elegant way?

I considered defining my own custom `DataLoader` that returns the simulated data that the sleep phase uses, but this started seeming like even more of a hack than the previous solution. After all, my "dataloader" doesn't load any data; it effectively generates new data every "epoch". It's seems unnatural to split the sleep phase updates in this way.

Is there a more straightforward way in lightning to train a network that doesn't require any data? Thanks!

**Comment by awaelchli** [comment]
Why is it not possible for you to write a Dataset class that generates the data? You mention you already have code for generating the data, so why not wrap it into a Dataset class?

**Comment by jeff-regier** [comment]
I had been concerned that the Dataset class would shuffle the data unnecessarily, thus slowing down training. But an `IterableDataset` seems to work: the `__iter__` method is overridden to return an iterator that yields a whole batch at once. Maybe this is the best way.

**Comment by awaelchli** [comment]
No, that is simply not true. The Dataloader is doing the batching and shuffling, not the dataset. Besides, shuffling does not slow training down. If your dataset has an undefined or infinite length, use the IterableDataset, otherwise use the regular Dataset class. Once you have that, just pass it to the Dataloader. All of this is regular PyTorch :)
